### PR TITLE
Recommend hanging indent in NEP1

### PR DESCRIPTION
**Pull request by arnetheduck**
This PR modernises the NEP1 style guide to prefer hanging indent over vertial alignment for long code statements while still allowing alignment in legacy code.

The change is based on research and study of existing style guides for both braced and indented languages that have seen wide adoption as well as working with a large Nim codebase with several teams touching the same code regularly.

The research was done as part of due diligence leading up to [nph](https://github.com/arnetheduck/nph) which uses this style throughout.

There are several reasons why hanging indent works well for collaboration, good code practices and modern Nim features:

* as NEP1 itself points out, alignment causes unnecessary friction when refactoring, adding/removing items to lists and otherwise improving code style or due to the need for realignment - the new recommendation aligns NEP1 with itself
* When collaborating, alignment leads to unnecessary git conflicts and blame changes - with hanging indent, such conflicts are minimised.
* Vertical alignment pushes much of the code to the right where often there is little space - when using modern features such as generics where types may be composed of several (descriptively named) components, there is simply no more room for parameters or comments
* The space to the left of the alignemnt cannot productively be used for anything (unlike on the right, where comments may be placed)
* Double hanging indent maintaines visual separation between parameters / condition and the body that follows.

This may seem like a drastic change, but in reality, it is not:

* the most popular editor for Nim (vscode) already promotes this style by default (if you press enter after `(`, it will jump to an indent on the next line)
* although orthogonal to these changes, tools such as `nph` can be used to reformat existing code should this be desired - when done in a single commit, `git blame` is not lost and neither are exsting PRs (they can simply be reformatted deterministically) - `nph` is also integrated with vscode.
* It only affects long lines - ie most code remains unchanged

Examples of vertical alignment in the wild, for wildly successful languages and formatters:

* [PEP-8](https://peps.python.org/pep-0008/#indentation) 
* [black](https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html#how-black-wraps-lines)
* [prettier](https://prettier.io/docs/en/)

The above examples are useful mainly to show that hanging-indent _generally_ is no impediment to efficient code reading and on the whole is an uncontroversial choice as befits the standard library.

### CODE REVIEW

**Comment by Araq** [comment]
Meh, it's not the style we use and optimizing for "git diffs" is fundamentally flawed as git diff doesn't understand ASTs and it encourages people to use as little indentation as possible ("early returns", yada yada) which makes me wonder why people don't just use assembler instead.

**Comment by arnetheduck** [comment]
> it's not the style we use

nep1:
```
Furthermore, just as
[Python's style guide](http://legacy.python.org/dev/peps/pep-0008/) changes
over time, this style guide will too
```

> encourages people to use as little indentation as possible

this is not covered by the change really, but the new recommended style _promotes_ the use of indentation: when you use indentation in the body, you don't have room for vertical alignment of function call parameters _and_ descriptive variable names at the same time. We hit this problem regularly in our codebase and hanging indent is the answer.

It also avoids git diffs where they are pointless, ie in function parameter lists.

**Comment by Araq** [comment]
As a compromise, this would be acceptable:

 ```nim
    type
      LongTupleA = tuple[
        wordyTupleMemberOne: int, wordyTupleMemberTwo: string,
        wordyTupleMemberThree: float]

 ```

- Similarly, any procedure and procedure type declarations that are longer
  than one line should do the same thing. Double indent may be used to
  distinguish them from the body that follows - this applies to all constructs
  with a body (if, while, etc).

    ```nim
    type
      EventCallback = proc(
        timeReceived: Time, errorCode: int, event: Event,
        output: var string)

    proc lotsOfArguments(
        argOne: string, argTwo: int, argThree: float,
        argFour: proc(), argFive: bool): GenericType[int, string] {.heyLookALongPragma.} =
      discard
    ```

- Multi-line procedure calls should continue indented (like multi-line procedure declarations).

    ```nim
    startProcess(
      nimExecutable, currentDirectory, compilerArguments
      environment, processOptions)
    ```

**Comment by arnetheduck** [comment]
> As a compromise,

**Comment by Araq** [comment]
Arguably one of the points of indentation based syntax is to avoid these lines with closing parenthesis as they add no value (of course the nested thing is closed eventually, that's enforced by the compiler, don't waste screen space on it).

**Comment by arnetheduck** [comment]
> As a compromise,

sure - the only practical difference I see is the new line before `)` - we tend to use it for proc:s in particular because then we _usually_ don't have to line-break the return type and pragmas that follow - line breaks there are usually nasty / ugly. 

This is a minor point though, I'm happy to change this PR as its focus is on the removal of alignment as an idea, not on a particular closing style (I don't think the closing style needs to be specified tbh, it's such a minor point that a human formatter, since these apparently still exist in 2023, can be left to decide on their own - I'm happy to change the examples).

**Comment by ASVIEST** [comment]
As for me best style is like that:
```nim
proc lotsOfArguments(
   argOne: string, argTwo: int, argThree: float,
   argFour: proc(), argFive: bool
): GenericType[int, string] {.heyLookALongPragma.} =
  discard

var flag = false
if (
  theLongCond1 and
  theLongCond2 and
  theLongCond3
): flag = True

if (
  theLongCond1 and
  theLongCond1
):
  echo "hi"
  flag = true

block checkFlag: #instead of one giant unreadable if
  #it can be better with unnamed breaks, but it deprecated 
  if not theVeryVeryVeryVeryTooLongCond:
    break checkFlag

  if not (
    theLongCond1 and
    theLongCond1
  ): break checkFlag

  #if body
  echo "hi"
  flag = True

case cond:
  of case1: something
  of case2: something2
```
I think ) should be on the next line because it can better show the return type with the pragma.

**Comment by Araq** [comment]
> As for me best style is like that ...

No, it's ugly and not negotiable.

**Comment by arnetheduck** [comment]
there - updated the simple cases, ie closing on same line as last entry - I left the return type on its own line since the line becomes very wide with it on the same line mostly to highlight it as an option - notably, it indents with the `proc` keyword to ensure maximum space for return type and pragma.

**Comment by metagn** [comment]
> I think ) should be on the next line because it can better show the return type with the pragma.

I think this is the only case where it's better, `):` being on its own line in `if` statements is hard to read. The visibility for the return type is important, there also doesn't seem to be a suggested style for when the return type is too long, only the pragma.

**Comment by arnetheduck** [comment]
> The visibility for the return type is important, there also doesn't seem to be a suggested style for when the return type is too long, only the pragma.

Well, a proc has 4 "expandable" sections: generics, params, return and pragma. So what we roughly end up with in a "max-complexity" scenario with the hanging indent style is name+`[` on one line, generics then `](` + params on one line, `):` + return + `{.` on one line and pragma list on the next (though usually on the same as return). 

There are a few ways to cut the above sections, but one way that works for all 4 is to put the "closing" on a new line and start the next expandable there - through experiments, I've found this works pretty well - ie it looks nicer to put return + pragma on a line on its own instead of params+return on one line then pragma on the next:

```nim
# force return on same line as params - hard to find it in the param noise
proc mylongfunctionname[A, B](
    paramOne: int, paramtwo: int, param3: SomeGeneric[int]): ReturnType {.
    pragma.} =
  discard

# move to the next line
proc mylongfunctionname[A: B](
    paramOne: int, paramtwo: int, param3: SomeGeneric[int]
): ReturnType {.pragma.} =
  discard
```

This point applies regardless if there's a pragma or not and we can "mostly" assume the pragma is low-information so having it more to the right and combined / after return is fine.

The moment we move parameters to a new line, putting the return type on the same line makes it "hard to find" among the parameter noise - notably, both styles occupy the same amount of lines but in the latter, we've emphasized the structure of the proc and its return.

This is more or less the algorithm that nph uses: if the 4 sections don't fit on a single line, move to a 3-line setup where the return comes after `):` on a new line - expressed as a scoring problem, line breaks within a section are more expensive than line breaks between sections so prefer the latter.

Whenever there's a complex return and just a few params, I've found that the 3-line style works well too, ie it's a fairly rare thing but here's a typical example:

```nim
proc someFunctionName(
    paramOne: int, paramTwo: float
): Future[Result[MygenericType[int], SomeError]] =
  discard

# We can keep playing this game and the style keeps working ;) This is based on a real-world
# example which looks silly here but makes some sense in its context
proc someFunctionName(
    paramOne: int, paramTwo: float
): Future[
    tuple[Result[MygenericType[int], SomeError], ExtraInformation],
] {.async: (raises: []).} =
  discard
```

There exist two-ine formattings for above that are decent but they indeed tend not to have the nice property of return-on-its-own.

When formatting manually, specially procs that don't have a return type, it's indeed easier to put `) =` on the same line as the last parameter, ie there's limited advantage in spending another line on that proc declaration.

If desired, the above style can be carried over to `if` also, creating a kind of consistency in the formatting:
```nim
if (
    someComplexConditions .. 
):
  body
```

I'm not advocating for the long-if style in this PR, only mentioning it as "room-for-growth", ie that hanging indent has ways to deal with the if as well that aren't terrible. If you think about it, each of `(` and `)` could be placed on a line of its own and there are styles that do so / people that prefer that - `(` on same line as previous token and `):` on its own is a "middle ground" option that people from both sides of that debate can at least accept as a compromise - it usually works well combined with "hugging", ie `call([\n...\n)]` for a call with multiple brackets.

Ok, that was a bit of a rant and detour into the deep end since I had to think about this recently - why that detour is worth making is mostly as a thought experiment to ensure that whatever style is picked has a reasonable behavior when expanded to more complex declarations and longer names - for example, in our code we often constrain generics which makes the "generic" section explode in length - if a working version of concepts make it to the language, this will become more common - the above strategy deals with this gracefully.

That said, in the vast majority of `proc`s it won't matter since everything typically fits on one line.

Long story short, I think flexibility on `)` placement is fine - it has little overall effect on formatting and placing it whereever can be made to work even when things go more complex - hanging indent, which this issue is about, however has a lot of objective advantages when applied to more complex code because it is more efficient at using the space available on the left that otherwise ends up being empty dead space.

Circling back to this PR specifically, the vertical alignment style fails the room-for-expansion test so here we are ;). 

NEP1 (just like large parts of stdlib) imo hasn't evolved with the language. The hanging indent change is _needed_ to continue improving the code, even if it doesn't document the "current style used".
